### PR TITLE
KAFKA-8993: EOS perf test framework

### PR DESCRIPTION
**Pull request by abbccdda**
Adapted from ProducerPerformance framework to test EOS application performance in a consume-process-produce loop.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Review by cpettitt-confluent — COMMENTED** [review]
Could we just add another option in ProducerPerformance to consumer from a topic? It already has consume from a file and generate in memory.  The benefit is that we have one place for all of the configurations and we have feature parity (e.g. throttling and whatever else is in producer performance).

File: tools/src/main/java/org/apache/kafka/tools/ClientTestHandler.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+ * limitations under the License.
+ */
+package org.apache.kafka.tools;
+
+import org.apache.kafka.clients.producer.ProducerRecord;
+
+public interface ClientTestHandler<K, V> {
+
+    /**
+     * Get one record for producing. Return null if no more record is left
```
Comment (cpettitt-confluent) [inline tools/src/main/java/org/apache/kafka/tools/ClientTestHandler.java]: Return null if no more *records are* left

File: tools/src/main/java/org/apache/kafka/tools/EosClientTestHandler.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        String applicationId = "Customized-EOS-Performance-" + UUID.randomUUID();
+        Properties consumerProperties = generateProps(consumerProps, consumerConfigFile);
+        consumerProperties.put(ConsumerConfig.GROUP_ID_CONFIG, applicationId);
+        consumerProperties.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
+        consumerProperties.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, "org.apache.kafka.common.serialization.ByteArrayDeserializer");
+        consumerProperties.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, "org.apache.kafka.common.serialization.ByteArrayDeserializer");
+        consumer = new KafkaConsumer<>(consumerProperties);
+        consumer.subscribe(Collections.singletonList(inputTopic));
+
+        recordsToBeSent = new LinkedList<>();
```
Comment (cpettitt-confluent) [inline tools/src/main/java/org/apache/kafka/tools/EosClientTestHandler.java]: Since we're performance testing it would be preferable to use ArrayList here. At the very least we do not have to do allocations per add. Also arrays as a primitive are much nicer for the CPU because you can pack them into cache lines vs. taking a cache miss on each iteration.

File: tools/src/main/java/org/apache/kafka/tools/ProducerTestHandler.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+class ProducerTestHandler implements ClientTestHandler<byte[], byte[]> {
+
+    private final Integer recordSize;
+    private final String payloadFilePath;
+    private final String payloadDelimiter;
+    private final List<byte[]> payloadByteList;
+    private final String outputTopic;
+    private byte[] payload;
+    private Random random;
```
Comment (cpettitt-confluent) [inline tools/src/main/java/org/apache/kafka/tools/ProducerTestHandler.java]: This should be final.

File: tools/src/main/java/org/apache/kafka/tools/ProducerTestHandler.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    public ProducerRecord<byte[], byte[]> getRecord() {
+        if (payloadFilePath != null) {
+            payload = payloadByteList.get(random.nextInt(payloadByteList.size()));
+        }
+        return new ProducerRecord<>(outputTopic, payload);
+    }
+
+    @Override
+    public void commit() {
+        // Nothing to be done
```
Comment (cpettitt-confluent) [inline tools/src/main/java/org/apache/kafka/tools/ProducerTestHandler.java]: Why not? We should be able to run with commit on the producer performance.

File: tools/src/main/java/org/apache/kafka/tools/KafkaClientPerformance.java [hunk 5]
```diff
@@ -274,28 +291,54 @@ private static ArgumentParser argParser() {
...
 
         parser.addArgument("--transactional-id")
-               .action(store())
-               .required(false)
-               .type(String.class)
-               .metavar("TRANSACTIONAL-ID")
-               .dest("transactionalId")
-               .setDefault("performance-producer-default-transactional-id")
-               .help("The transactionalId to use if transaction-duration-ms is > 0. Useful when testing the performance of concurrent transactions.");
+                .action(store())
```
Comment (cpettitt-confluent) [inline tools/src/main/java/org/apache/kafka/tools/KafkaClientPerformance.java:309]: I think we need to remove the leading space here and on the next three sections? It would be good to not change the line of code if the only difference is whitespace. It will also keep your name out of `git blame` unnecessarily :).

File: tools/src/main/java/org/apache/kafka/tools/EosClientTestHandler.java [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+import java.time.Duration;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Properties;
+import java.util.UUID;
+
+import static org.apache.kafka.tools.KafkaClientPerformance.generateProps;
+
+class EosClientTestHandler<K, V> implements ClientTestHandler<K, V> {
```
Comment (cpettitt-confluent) [inline tools/src/main/java/org/apache/kafka/tools/EosClientTestHandler.java]: Maybe ConsumerAndProducerTestHandler? We should be able to test without transactions.

File: tools/src/main/java/org/apache/kafka/tools/EosClientTestHandler.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+    @Override
+    public ProducerRecord<K, V> getRecord() {
+        if (recordsToBeSent.isEmpty()) {
+            ConsumerRecords<K, V> consumedRecords = consumer.poll(Duration.ofMillis(pollTimeout));
+            for (ConsumerRecord<K, V> consumerRecord : consumedRecords) {
+                recordsToBeSent.add(new ProducerRecord<>(outputTopic, consumerRecord.value()));
+            }
+        }
+        return recordsToBeSent.remove(0);
```
Comment (cpettitt-confluent) [inline tools/src/main/java/org/apache/kafka/tools/EosClientTestHandler.java]: So, we should probably use an ArrayDeque actually. I took a quick peek at its implementation and it uses a circular buffer, which should give us the benefits of ArrayList mentioned above while also preserving order and keeping remove from the front at O(1).

File: tools/src/main/java/org/apache/kafka/tools/EosClientTestHandler.java [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+    }
+
+    @Override
+    public void commit() {
+        consumer.commitSync();
```
Comment (cpettitt-confluent) [inline tools/src/main/java/org/apache/kafka/tools/EosClientTestHandler.java]: We should be able to choose whether we're using transactions or not, even for this test?

**Review by cpettitt-confluent — COMMENTED** [review]
This looks pretty good. Some specifics in comments.

File: tools/src/main/java/org/apache/kafka/tools/KafkaClientPerformance.java (hunk 5 above)
Comment (abbccdda) [reply tools/src/main/java/org/apache/kafka/tools/KafkaClientPerformance.java:309]: Lol you remind me of the Linkedin style.

File: tools/src/main/java/org/apache/kafka/tools/EosClientTestHandler.java (hunk 2 above)
Comment (abbccdda) [reply tools/src/main/java/org/apache/kafka/tools/EosClientTestHandler.java]: The reason I do linked list is because we need to `remove` node everytime we populate a new producer record. What do you think would be a good trade-off data structure which supports 1. batch load-in 2. pop at the end

File: tools/src/main/java/org/apache/kafka/tools/EosClientTestHandler.java (hunk 6 above)
Comment (abbccdda) [reply tools/src/main/java/org/apache/kafka/tools/EosClientTestHandler.java]: I guess a ConsumerTestHandler is better, as this doesn't have anything to do with producer.

File: tools/src/main/java/org/apache/kafka/tools/KafkaClientPerformance.java (hunk 5 above)
Comment (abbccdda) [reply tools/src/main/java/org/apache/kafka/tools/KafkaClientPerformance.java:309]: The thing is that this indentation itself is not correct, it's neither matching an indentation number (15 instead of 16), nor aligning with `.add`. So I personally feel this is not a big ideal, but could revert this if you feel strong about this.

File: tools/src/main/java/org/apache/kafka/tools/EosClientTestHandler.java (hunk 2 above)
Comment (cpettitt-confluent) [reply tools/src/main/java/org/apache/kafka/tools/EosClientTestHandler.java]: See 4 comments down about ArrayDeque.

File: tools/src/main/java/org/apache/kafka/tools/EosClientTestHandler.java (hunk 7 above)
Comment (abbccdda) [reply tools/src/main/java/org/apache/kafka/tools/EosClientTestHandler.java]: Sg!

File: tools/src/main/java/org/apache/kafka/tools/KafkaClientPerformance.java [hunk 9]
```diff
@@ -52,72 +48,47 @@ public static void main(String[] args) throws Exception {
...
-                }
-
+            Properties props = generateProps(producerProps, producerConfig);
             props.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, "org.apache.kafka.common.serialization.ByteArraySerializer");
             props.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, "org.apache.kafka.common.serialization.ByteArraySerializer");
+
+            final String applicationId = "Customized-EOS-Performance-" + UUID.randomUUID();
             if (transactionsEnabled)
-                props.put(ProducerConfig.TRANSACTIONAL_ID_CONFIG, transactionalId);
+                props.put(ProducerConfig.TRANSACTIONAL_ID_CONFIG, applicationId);
```
Comment (cpettitt-confluent) [inline tools/src/main/java/org/apache/kafka/tools/KafkaClientPerformance.java]: This makes the transactionalId non-deterministic. Is there a good reason to do that? I think we should always blow away the destination topic, so if we ever end up using the same transactionalId that would indicate a problem in our test setup?

File: tools/src/main/java/org/apache/kafka/tools/KafkaClientPerformance.java [hunk 10]
```diff
@@ -126,28 +97,30 @@ public static void main(String[] args) throws Exception {
...
                     producer.beginTransaction();
                     transactionStartTime = System.currentTimeMillis();
                 }
 
-
-                if (payloadFilePath != null) {
-                    payload = payloadByteList.get(random.nextInt(payloadByteList.size()));
-                }
-                record = new ProducerRecord<>(topicName, payload);
+                record = clientTestHandler.getRecord();
```
Comment (cpettitt-confluent) [inline tools/src/main/java/org/apache/kafka/tools/KafkaClientPerformance.java:110]: Since it looks like we expect numRecords to always be set, should we error out here if record is null? That would seem to indicate that our source topic was under-sized or some other issue (GC pause?) caused poll to timeout.

File: tools/src/main/java/org/apache/kafka/tools/KafkaClientPerformance.java (hunk 9 above)
Comment (abbccdda) [reply tools/src/main/java/org/apache/kafka/tools/KafkaClientPerformance.java]: I see, the challenge here is that whether it's worth blowing away the input topic every time, as now the producer will send offsets for the progress tracking, which means for next test job to be run we need to clean up the old offset.

Comment (cpettitt-confluent) [reply tools/src/main/java/org/apache/kafka/tools/KafkaClientPerformance.java]: How about we reset the offsets between tests? I think that may be what you're suggesting. It would be cleaner not to accumulate metadata over time.

Comment (abbccdda) [reply tools/src/main/java/org/apache/kafka/tools/KafkaClientPerformance.java]: Sg

Comment (abbccdda) [reply tools/src/main/java/org/apache/kafka/tools/KafkaClientPerformance.java]: While I was doing the work, I realized that our intent to delete/reset offset could fail. I still updated code to do offset reset as it's more metadata size bounded, just bring this up and see you feel about the two ways here.

Comment (cpettitt-confluent) [reply tools/src/main/java/org/apache/kafka/tools/KafkaClientPerformance.java]: I've been doing the delete / reset offset path in all of my tests and it seems to be going fine. For the output topic I delete and do create in a loop because the list command sometimes does not show the topic but then create fails because it thinks it still exists. We could get the committed offset at the beginning and fail if it is greater than 0.

File: tools/src/main/java/org/apache/kafka/tools/KafkaClientPerformance.java (hunk 10 above)
Comment (abbccdda) [reply tools/src/main/java/org/apache/kafka/tools/KafkaClientPerformance.java:110]: I added a null check to be safe here.

File: tools/src/main/java/org/apache/kafka/tools/KafkaClientPerformance.java (hunk 9 above)
Comment (abbccdda) [reply tools/src/main/java/org/apache/kafka/tools/KafkaClientPerformance.java]: Sg, I finally used a `seek` call to reset offset by the end.

File: docs/streams/developer-guide/processor-api.html [hunk 11]
```diff
@@ -361,7 +361,7 @@ <h2>
             <h2><a class="toc-backref" href="#id10">Accessing Processor Context</a><a class="headerlink" href="#accessing-processor-context" title="Permalink to this headline"></a></h2>
             <p>As we have mentioned in the <a href=#defining-a-stream-processor>Defining a Stream Processor</a> section, a <code>ProcessorContext</code> control the processing workflow, such as scheduling a punctuation function, and committing the current processed state.</p>
             <p>This object can also be used to access the metadata related with the application like
-            <code class="docutils literal"><span class="pre">applicationId</span></code>, <code class="docutils literal"><span class="pre">taskId</span></code>,
```
Comment (cpettitt-confluent) [inline docs/streams/developer-guide/processor-api.html]: This looks like it should go in regardless of the perf tool work. I'd suggest moving this to a separate PR.

File: docs/streams/upgrade-guide.html [hunk 12]
```diff
@@ -813,7 +813,7 @@ <h3><a id="streams_api_changes_0101" href="#streams_api_changes_0101">Streams AP
 
     <p> TopologyBuilder: </p>
     <ul>
-        <li> methods #sourceTopics(String applicationId) and #topicGroups(String applicationId) got simplified to #sourceTopics() and #topicGroups() </li>
```
Comment (cpettitt-confluent) [inline docs/streams/upgrade-guide.html]: Same comment as above

File: tools/src/main/java/org/apache/kafka/tools/ConsumerTestHandler.java [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+
+/**
+ * A consumer based EOS test handler that will poll data from input topic and
+ * buffer them in an in-memory queue.
+ */
+class ConsumerTestHandler<K, V> implements ClientTestHandler<K, V> {
+
+    private final Consumer<K, V> consumer;
+    private final String consumerGroupId;
+    private Deque<ProducerRecord<K, V>> recordsToBeSent;
```
Comment (cpettitt-confluent) [inline tools/src/main/java/org/apache/kafka/tools/ConsumerTestHandler.java]: Can be final

File: tools/src/main/java/org/apache/kafka/tools/ConsumerTestHandler.java [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+
+        this.producer = producer;
+        recordsToBeSent = new ArrayDeque<>();
+    }
+
+    @Override
+    public ProducerRecord<K, V> getRecord() {
+        System.out.println("Consumer polls new data with timeout " + pollTimeout);
+        while (recordsToBeSent.isEmpty()) {
+            ConsumerRecords<K, V> consumedRecords = consumer.poll(Duration.ofMillis(pollTimeout));
```
Comment (cpettitt-confluent) [inline tools/src/main/java/org/apache/kafka/tools/ConsumerTestHandler.java:90]: It looks like my comment below about getting a null message (as mentioned in the interface doc) would not happen because we'd spin here forever. Probably we should return null here if the fetch fails? Then in the loop below we may want to either fail immediately, because the test is going to be skewed or at least fail after n retries to avoid hanging everything, e.g. in the case of an undersized input topic.

File: tools/src/main/java/org/apache/kafka/tools/KafkaClientPerformance.java [hunk 15]
```diff
@@ -156,6 +136,10 @@ public static void main(String[] args) throws Exception {
             if (transactionsEnabled && currentTransactionSize != 0)
                 producer.commitTransaction();
 
+            if (flushEnabled) {
+                producer.flush();
```
Comment (cpettitt-confluent) [inline tools/src/main/java/org/apache/kafka/tools/KafkaClientPerformance.java:142]: clientTestHandler.onFlush()?

File: tools/src/main/java/org/apache/kafka/tools/KafkaClientPerformance.java [hunk 16]
```diff
@@ -156,6 +136,10 @@ public static void main(String[] args) throws Exception {
             if (transactionsEnabled && currentTransactionSize != 0)
                 producer.commitTransaction();
```
Comment (cpettitt-confluent) [inline tools/src/main/java/org/apache/kafka/tools/KafkaClientPerformance.java:139]: clientTestHandler.onTxnCommit?

File: tools/src/main/java/org/apache/kafka/tools/ClientTestHandler.java [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+public interface ClientTestHandler<K, V> {
+
+    /**
+     * Get one record to produce. Return null if no more records are left.
+     */
+    ProducerRecord<K, V> getRecord();
+
+    void onFlush();
+
+    void onTxnCommit();
```
Comment (cpettitt-confluent) [inline tools/src/main/java/org/apache/kafka/tools/ClientTestHandler.java]: I would suggest renaming these to flush / txnCommit or even flush / commit as these do the actual work.

File: tools/src/main/java/org/apache/kafka/tools/ClientTestHandler.java [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.tools;
+
+import org.apache.kafka.clients.producer.ProducerRecord;
+
+public interface ClientTestHandler<K, V> {
```
Comment (cpettitt-confluent) [inline tools/src/main/java/org/apache/kafka/tools/ClientTestHandler.java:24]: It's worth adding a brief comment to this as it may show up in searches. Something as simple as "An object used by {@link KafkaClientPerformance} to test the performance of Kafka clients".

**Review by cpettitt-confluent — COMMENTED** [review]
Looks close to mergeable. I've added a few comments that should be addressed or commented on. I would also like to get this running at least once in a test environment to verify it works. If you've already done that great. If you need help doing that, let me know.

File: docs/streams/developer-guide/processor-api.html (hunk 11 above)
Comment (abbccdda) [reply docs/streams/developer-guide/processor-api.html]: sg

File: tools/src/main/java/org/apache/kafka/tools/ConsumerTestHandler.java (hunk 14 above)
Comment (abbccdda) [reply tools/src/main/java/org/apache/kafka/tools/ConsumerTestHandler.java:90]: Yes, that's a good catch!

File: tools/src/main/java/org/apache/kafka/tools/ConsumerTestHandler.java [hunk 19]
```diff
@@ -0,0 +1,134 @@
...
+        List<String> consumerProps = res.getList("consumerConfig");
+        String consumerConfigFile = res.getString("consumerConfigFile");
+        pollTimeout = res.getLong("pollTimeout");
+        inputTopic = res.getString("inputTopic");
+        outputTopic = res.getString("topic");
+
+        Properties consumerProperties = KafkaClientPerformance.generateProps(consumerProps, consumerConfigFile);
+        consumerProperties.put(ConsumerConfig.GROUP_ID_CONFIG, consumerGroupId);
+        consumerProperties.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
+        consumerProperties.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, "org.apache.kafka.common.serialization.ByteArrayDeserializer");
```
Comment (cpettitt-confluent) [inline tools/src/main/java/org/apache/kafka/tools/ConsumerTestHandler.java]: Nit: I know this was inherited from the existing code, but it would be nice to do something like `ByteArrayDeserializer.class.getName()` for serializers and deserializers.

File: tools/src/main/java/org/apache/kafka/tools/KafkaClientPerformance.java [hunk 20]
```diff
@@ -43,111 +38,96 @@
...
-            if (producerProps != null)
-                for (String prop : producerProps) {
-                    String[] pieces = prop.split("=");
-                    if (pieces.length != 2)
-                        throw new IllegalArgumentException("Invalid property: " + prop);
-                    props.put(pieces[0], pieces[1]);
-                }
-
+            Properties props = generateProps(producerProps, producerConfig);
             props.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, "org.apache.kafka.common.serialization.ByteArraySerializer");
```
Comment (cpettitt-confluent) [inline tools/src/main/java/org/apache/kafka/tools/KafkaClientPerformance.java]: Nit: If we do fix up the above example of this it makes sense to fix this up too.

File: tools/src/main/java/org/apache/kafka/tools/ClientTestHandler.java [hunk 21]
```diff
@@ -0,0 +1,36 @@
...
+
+    /**
+     * Get one record to produce. Return null if no more records are left.
+     */
+    ProducerRecord<K, V> getRecord();
+
+    void flush();
+
+    void txnCommit();
```
Comment (cpettitt-confluent) [inline tools/src/main/java/org/apache/kafka/tools/ClientTestHandler.java]: Nit: looking at this again, `commitTransaction` would be more consistent with the transaction API for the client.

**Review by cpettitt-confluent — APPROVED** [review]
Nit = optional fix, subjective, not required for merge.

All comments here are nits, so I'm OK with this going in as is, or with some / all nits fixed without further review. If anything more substantial changes, let's have another look.

File: tools/src/main/java/org/apache/kafka/tools/ConsumerTestHandler.java (hunk 19 above)
Comment (abbccdda) [reply tools/src/main/java/org/apache/kafka/tools/ConsumerTestHandler.java]: Sg!

File: tools/src/main/java/org/apache/kafka/tools/KafkaClientPerformance.java (hunk 20 above)
Comment (abbccdda) [reply tools/src/main/java/org/apache/kafka/tools/KafkaClientPerformance.java]: Yep

**Comment by abbccdda** [comment]
Fixing the last nit comments from Chirs, while noting that we got 2/3 green builds already with jdk 8 as a flaky test.

**Review by guozhangwang — COMMENTED** [review]
@abbccdda I did a quick search for `ProducerPerformance` on the codebase and there are a lot more matches that need to be updated. E.g. `producer_performance.py`, and `kafka-producer-perf-test.sh`.

**Comment by cpettitt-confluent** [comment]
My suggestion would be to move the new perf test code to the streams-perf repo for now. We need to do some adapting to the perf test tools anyway, and this also allows us to iterate more quickly, not break existing uses, etc.

**Comment by guozhangwang** [comment]
> My suggestion would be to move the new perf test code to the streams-perf repo for now. We need to do some adapting to the perf test tools anyway, and this also allows us to iterate more quickly, not break existing uses, etc.

Sounds good to me.

**Comment by abbccdda** [comment]
I would say we should attempt to merge at this point in parallel, as having a diverged codebase later on is hard to backport. We also don't want to maintain a stream perf repo which is completely out-of-sync and hard for other community folks to reproduce the test result we have, if they want to test on their end as well. If community couldn't reproduce, there is no point of maintaining this codepath in AK as well, since we are speaking different language from people outside of streams perf. From my past experience, diverged codebase is very hard to be maintained.

**Comment by cpettitt-confluent** [comment]
I'm fine merging all of the perf testing work we're doing back to AK after it has stabilized. In the short term trying to do this work in AK is going to slow things down significantly. Given the change in scope of the test tool, I think it is reasonable for them to diverge, which ProducerPerformance remaining as is, with a much narrower scope.

**Comment by abbccdda** [comment]
I don't think there is a huge amount of work to merge this PR. For example we just need to change the reference in `producer_performance.py` to let it adapt to the new interface, and `kafka-producer-perf-test. These combined shouldn't take more than one hour to finish. I will definitely prioritize the merging with perf test tool also, while in the meantime trying to make these changes verified as well.

**Comment by mjsax** [comment]
Closing this stale PR due to inactivity.
